### PR TITLE
Add ZStream#debounce

### PR DESCRIPTION
**Pull request by luis3m**
- Both JVM & JS tests passing on top of https://github.com/zio/zio/pull/3424 as it guarantees to run the effects in the correct order.

- Took two tests from https://github.com/zio/zio/pull/3521 (cc @reibitto) and added three more.

- It does reset time

### CODE REVIEW

**Comment by luis3m** [comment]
These two failures seem to be unrelated

File: streams/shared/src/main/scala/zio/stream/ZStream.scala [hunk 1]
```diff
@@ -2469,6 +2469,60 @@ abstract class ZStream[-R, +E, +O](val process: ZManaged[R, Nothing, ZIO[R, Opti
...
+    ZStream[R with Clock, E, O] {
+      for {
+        chunks <- self.process
+        ref    <- Ref.make[Option[Fiber[E, Chunk[O]]]](None).toManaged_
+        pull = {
+          ref.get.flatMap {
+            case None =>
+              chunks >>= store(ref)
+            case Some(previous) =>
+              val prev = previous.join.run.map(_.mapError(Some(_)))
```
Comment (iravid) [inline streams/shared/src/main/scala/zio/stream/ZStream.scala:2497]:
```suggestion
              val prev = previous.await.map(_.mapError(Some(_)))
```

File: streams/shared/src/main/scala/zio/stream/ZStream.scala [hunk 2]
```diff
@@ -2469,6 +2469,60 @@ abstract class ZStream[-R, +E, +O](val process: ZManaged[R, Nothing, ZIO[R, Opti
...
+                          fiber  <- loser.merge.join
+                          done   <- ZIO.done(fiber)
+                          _      <- ref.set(None)
+                          result <- Pull.emit(done)
+                        } yield result
+                    },
+                    chunk =>
+                      loser.fold(
+                        _.interrupt *> store(ref)(chunk),
+                        _.interrupt *> ref.set(None) *> Pull.emit(chunk)
```
Comment (iravid) [inline streams/shared/src/main/scala/zio/stream/ZStream.scala:2516]: We can't interrupt an ongoing pull. That is likely to lead to corruption of internal operator state. We have to keep the ongoing pull around.

Comment (luis3m) [reply streams/shared/src/main/scala/zio/stream/ZStream.scala]: @iravid solved

**Comment by luis3m** [comment]
1) Refactored a bit
2) Some extra annotations to fix Dotty compilation issues

**Comment by luis3m** [comment]
I can remove the top level type parameters by deferring them to `State` but I'm not sure it's worth doing
```scala
sealed abstract class State[+E, +O]
case object NotStarted                                      extends State[Nothing, Nothing]
case class Previous[O](fiber: Fiber[Nothing, Chunk[O]])     extends State[Nothing, O]
case class Current[E, O](fiber: Fiber[Option[E], Chunk[O]]) extends State[E, O]
case object Done                                            extends State[Nothing, Nothing]
```

**Comment by reibitto** [comment]
@iravid @luis3m Hmm, I think there's one corner case still left with the handling of the end of a stream. I expect the following to result in `List(3)`, but right now it's an empty list:

```scala
ZStream(1, 2, 3).fixed[Any](500.millis).debounce(1.second).runCollect
```

**Comment by iravid** [comment]
Ah ok. That should be fixed. Can you open a follow up issue?

**Comment by reibitto** [comment]
Sure thing.

**Pull request merged**